codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -243,66 +246,43 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
-
-            if (lastDataWrite != null && !endOfStream) {
-                throw new IllegalStateException(
-                        "Sending non-trailing headers after data has been sent for stream: "
-                                + streamId);
+            ChannelFuture future;
+            if (openStream) {
```

Comment:
@nmittler - Just to clarify what I think you are suggesting...have the flow controller just check if things are queued, then queue the frame, otherwise send immediately?  So the state management stays largely the same as it was in `DefaultHttp2ConnectionEncoder` before this PR?

I originally thought you were suggesting moving this code block verbatim (setting/checking the stream state) into the flow controller.